packages/libusb/termux-usb-support.patch

```diff
@@ -0,0 +1,173 @@
+ static int usbdev_names = 0;
+@@ -105,6 +107,10 @@
+ static int linux_default_scan_devices(struct libusb_context *ctx);
+ #endif
+ 
++#if defined(TERMUX)
++static int linux_termux_scan_devices(struct libusb_context *ctx);
++#endif
++
+ struct kernel_version {
+ 	int major;
+ 	int minor;
+@@ -127,6 +133,9 @@
+ 	size_t descriptors_len;
+ 	struct config_descriptor *config_descriptors;
+ 	int active_config; /* cache val for !sysfs_available  */
++#if defined(TERMUX)
++	int termux_fd;
++#endif
+ };
+ 
+ struct linux_device_handle_priv {
+@@ -188,6 +197,26 @@
+ 	char path[24];
+ 	int fd;
+ 
++#if defined(TERMUX)
++	struct linux_device_priv *priv = usbi_get_device_priv(dev);
++
++	if ((fd = priv->termux_fd) != -1) {
++		/* The returned fd (from this function) might gets closed at some
++		 * point in the future when the user close the handle
++		 * Avoid touching the fd given by Android since we might want to close
++		 * and reopen the device again */
++		if ((fd = dup(priv->termux_fd)) == -1) {
++			usbi_err(ctx, "unable to duplicate termux usb fd: %s", strerror(errno));
++			return LIBUSB_ERROR_IO;
++		}
++		/* Android gives fd with O_RDWR by default, and changing the mode
++		 * of an fd is not possible at the moment */
++		usbi_info(ctx, "setting fd mode for termux supplied usb device is unsupported, the fd will have mode O_RDWR instead");
++
++		return fd;
++	}
++#endif
++
+ 	if (usbdev_names)
+ 		snprintf(path, sizeof(path), USBDEV_PATH "/usbdev%u.%u",
+ 			dev->bus_number, dev->device_address);
+@@ -458,12 +487,20 @@
+ 
+ 	usbi_mutex_static_lock(&linux_hotplug_lock);
+ 
++#if defined(TERMUX)
++	if (getenv(TERMUX_USB_ENV)) {
++		ret = linux_termux_scan_devices(ctx);
++		goto out;
++	}
++#endif
++
+ #if defined(HAVE_LIBUDEV)
+ 	ret = linux_udev_scan_devices(ctx);
+ #else
+ 	ret = linux_default_scan_devices(ctx);
+ #endif
+ 
++out:
+ 	usbi_mutex_static_unlock(&linux_hotplug_lock);
+ 
+ 	return ret;
+@@ -916,6 +953,10 @@
+ 	int fd, speed, r;
+ 	ssize_t nb;
+ 
++#if defined(TERMUX)
++	priv->termux_fd = -1;
++#endif
++
+ 	dev->bus_number = busnum;
+ 	dev->device_address = devaddr;
+ 
+@@ -1328,6 +1369,74 @@
+ }
+ #endif
+ 
++#if defined(TERMUX)
++static int linux_termux_get_devinfo(int fd, uint8_t *busnum, uint8_t *devaddr)
++{
++	char proc_path[PATH_MAX], usbfs_path[PATH_MAX];
++
++	snprintf(proc_path, PATH_MAX, "/proc/self/fd/%u", fd);
++	if (readlink(proc_path, usbfs_path, PATH_MAX) < 0 ||
++		strncmp("/dev/bus/usb/", usbfs_path, 12)) {
++		return LIBUSB_ERROR_IO;
++	}
++
++	if (sscanf(usbfs_path, "/dev/bus/usb/%hhu/%hhu", busnum, devaddr) != 2)
++		return LIBUSB_ERROR_IO;
++
++	return LIBUSB_SUCCESS;
++}
++
++static int linux_termux_scan_devices(struct libusb_context *ctx)
++{
++	struct libusb_device *dev;
++	struct linux_device_priv *priv;
++	uint8_t busnum, devaddr;
++	unsigned long session_id;
++	char *envval, *envval_end;
++	int fd, r;
++
++	assert((envval = getenv(TERMUX_USB_ENV)));
```

Comment:
the assert and assignment should be separate, just in case someone tries to build it with -NDEBUG

Reply:
Right, quite an oversight that was.